iib/workers/tasks/build.py

```diff
@@ -337,6 +342,166 @@ def _get_resolved_image(pull_spec):
+    :param str base_dir: base directory to which the database file should be saved.
+    :return: path to the copied database file.
+    :rtype: str
+    :raises IIBError: if any podman command fails.
+    """
+    data = skopeo_inspect(f'docker://{from_index}')
+    try:
+        db_path = data['Labels']['operators.operatorframework.io.index.database.v1']
+    except KeyError:
+        raise IIBError('Index image doesn\'t have the label specifying its database location.')
+    _copy_files_from_image(from_index, db_path, base_dir)
+    local_path = os.path.join(base_dir, os.path.basename(db_path))
+    return local_path
+
+
+def _serve_index_registry(db_path):
+    """
+    Locally start OPM registry service, which can be communicated with using gRPC queries.
+
+    Due to IIB's paralellism, the service can run multiple times, which could lead to port
+    binding conflicts. Resolution of port conflicts is handled in this function as well.
+
+    :param str db_path: path to index database containing the registry data.
+    :return: tuple containing port number of the running service and the running Popen object.
+    :rtype: (int, Popen)
+    :raises IIBError: if all tried ports are in use, or the command failed for another reason.
+    """
+    conf = get_worker_config()
+    port_start = conf['iib_grpc_start_port']
+    port_end = port_start + conf['iib_grpc_max_port_tries']
+
+    for port in range(port_start, port_end):
+        try:
+            return (
+                port,
+                _serve_index_registry_at_port(
+                    db_path, port, conf['iib_grpc_max_tries'], conf['iib_grpc_init_wait_time']
+                ),
+            )
+        except AddressAlreadyInUse:
+            log.info('Port %d is in use, trying another.', port)
+
+    err_msg = (
+        'No free port has been found after ' f'{conf.get("iib_grpc_max_port_tries")} attempts.'
+    )
+    log.error(err_msg)
+    raise IIBError(err_msg)
+
+
+def _serve_index_registry_at_port(db_path, port, max_tries, wait_time):
+    """
+    Start an image registry service at a specified port.
+
+    :param str db_path: path to index database containing the registry data.
+    :param str int port: port to start the service on.
+    :param max_tries: how many times to try to start the service before giving up.
+    :param wait_time: time to wait before checking if the service is initialized.
+    :return: object of the running Popen process.
+    :rtype: Popen
+    :raises IIBError: if te process has failed to initialize too many times, or an unexpected
+        error occured.
+    :raises AddressAlreadyInUse: if the specified port is already being used by another service.
+    """
+    cmd = ['opm', 'registry', 'serve', '-p', str(port), '-d', db_path]
+    for attempt in range(max_tries):
+        rpc_proc = subprocess.Popen(
+            cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True
+        )
+        time.sleep(wait_time)
```

Comment:
The problem I most frequently encountered when prototyping was that ocp would arbitrarily not output any messages at all, requiring a restart. This is why I opted for restarting the service multiple times instead of retrying for the output.

I suppose I could add the querying mechanism if we expect  a very high variability in start-up time (If we expect the service start up time to sometimes take 10+ seconds then the passive wait time would be inefficient), but I haven't encountered this problem during prototyping. Then again, I didn't really stress test it. 

Thoughts?

Reply:
I feel strongly about not relying on sleeps.

Reply:
I added a loop to check program output until timeout. But I had to add a one second sleep between attempts because otherwise the program didn't give me any output for some reason. Also, I feel that adding CPU intensive active loops is not a good idea. 

I increased the default timeout to 3 seconds so that it would loop at least a couple times and because I don't know the precision of time provided by the system.

Reply:
Sorry, I should be clearer. I feel strongly about not relying on an arbitrary sleep time to hope that it was long enough. Using sleeps to create small delays between checks is totally fine.